File: 2015/6/15Day6.py
import re
import numpy as np
import os

# Define the directory and file name
directory = r'C:\Users\User\Documents\AdventofCode\2015\6'

D6_file = 'Day6_lights.txt'
D6_file_path = os.path.join(directory, D6_file)

with open(D6_file_path) as file:
    lights_inst = file.read()

lights_inst = lights_inst.splitlines()

def parse_instruction(instruction):
    # Regex to capture the action and the four numbers (coordinates)
    match = re.match(r"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)", instruction)
    
    if match:
        action = match.group(1)
        x1 = int(match.group(2))
        y1 = int(match.group(3))
        x2 = int(match.group(4))
        y2 = int(match.group(5))
        
        # Return in the required format
        return [action, x1, y1, x2, y2]
    else:
        return None

instruction_matrix = []

for n in range(len(lights_inst)):
    instruction_n = lights_inst[n]
    instruction_table = parse_instruction(instruction_n)
    instruction_matrix.append(instruction_table)
instruction_matrix = np.array(instruction_matrix)


# Create a binary matrix to represent all houes with lights 
# All lights start turned off(0)
light_matrix = np.zeros((1000, 1000)) 

for n in range(len(lights_inst)):
    instruction_n = instruction_matrix[n]
    
    type = instruction_n[0]    
    if type == "turn on":
        action = 1
    elif type == "turn off":
        action = 0
    else:
        action = -1  # Need to absolute it
    
    x1 = int(instruction_n[1])
    y1 = int(instruction_n[2])
    x2 = int(instruction_n[3])
    y2 = int(instruction_n[4])
    print(f"x={x2-x1},y={y2-y1}")
    
    
print(D6_file_path)
    

# toggle   = -1 and then absolute
# turn on  == 1
# turn off == 0

File: 2015/18/15Day6.rb
require 'matrix'

D6_file = 'Day6_lights.txt'
D6_file_path = File.join(File.dirname(File.expand_path(__FILE__)), D6_file)

lights_inst = File.read(D6_file_path).split("\n")

def parse_instruction(instruction)
  # Regex to capture the action and the four numbers (coordinates)
  match = instruction.match(/(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)/)

  if match
    action = match[1]
    x1 = match[2].to_i
    y1 = match[3].to_i
    x2 = match[4].to_i
    y2 = match[5].to_i

    # Return in the required format
    [action, x1, y1, x2, y2]
  else
    nil
  end
end

instruction_matrix = []

lights_inst.each do |instruction_n|
  instruction_matrix << parse_instruction(instruction_n)
end

# Create a 1000x1000 matrix for light status
light_matrix_1 = Array.new(1000) { Array.new(1000, 0) }

instruction_matrix.each do |instruction_n|
  action = instruction_n[0]
  x1 = instruction_n[1]
  y1 = instruction_n[2]
  x2 = instruction_n[3]
  y2 = instruction_n[4]

  (x1..x2).each do |x|
    (y1..y2).each do |y|
      case action
      when "turn on"
        light_matrix_1[x][y] = 1
      when "turn off"
        light_matrix_1[x][y] = 0
      when "toggle"
        light_matrix_1[x][y] = 1 - light_matrix_1[x][y] # Toggle the light
      end
    end
  end
end

total_sum_1 = light_matrix_1.flatten.sum
puts "Total houses with lights on: #{total_sum_1}"

# Create a 1000x1000 matrix for brightness
light_matrix_2 = Array.new(1000) { Array.new(1000, 0) }

instruction_matrix.each do |instruction_n|
  action = instruction_n[0]
  x1 = instruction_n[1]
  y1 = instruction_n[2]
  x2 = instruction_n[3]
  y2 = instruction_n[4]

  (x1..x2).each do |x|
    (y1..y2).each do |y|
      case action
      when "turn on"
        light_matrix_2[x][y] += 1
      when "turn off"
        light_matrix_2[x][y] -= 1
        light_matrix_2[x][y] = 0 if light_matrix_2[x][y] < 0
      when "toggle"
        light_matrix_2[x][y] += 2
      end
    end
  end
end

total_sum_2 = light_matrix_2.flatten.sum
puts "Total brightness of houses: #{total_sum_2}"
